src/icu_win.cc

```diff
@@ -0,0 +1,238 @@
+    dynamic_buffer_size = result;
+  }
+  // Slow-path. Allocate a buffer of the required size.
+  std::unique_ptr<wchar_t[]> buffer(new wchar_t[dynamic_buffer_size]);
+  const UINT length = ::GetSystemDirectoryW(buffer.get(), dynamic_buffer_size);
+  if (length <= dynamic_buffer_size) {
+    return std::wstring(buffer.get(), length);
+  }
+  return std::wstring();
+}
+
+IcuFunctions GetIcuFunctions() {
+  if (g_unavailable.load(std::memory_order_relaxed)) {
+    return Unavailable();
+  }
+
+  // Check if already loaded
+  {
+    const auto ucal_getHostTimeZoneRef =
+        g_ucal_getHostTimeZone.load(std::memory_order_relaxed);
+    const auto ucal_getTimeZoneIDForWindowsIDRef =
+        g_ucal_getTimeZoneIDForWindowsID.load(std::memory_order_relaxed);
+
+    // Note: ucal_getHostTimeZone can be unavailable on older Windows.
+    if (ucal_getTimeZoneIDForWindowsIDRef != nullptr) {
+      return {true, ucal_getHostTimeZoneRef, ucal_getTimeZoneIDForWindowsIDRef};
+    }
+  }
+
+  // Our goal here is to load the ICU DLL from the system directory, even when
+  // the current process has loaded "icu.dll" from somewhere other than the
+  // system directory. To do so we must use the full path here.
+  const std::wstring icu_dll_path = GetSystem32Dir() + L"\\icu.dll";
+
+  // CAVEAT: LoadLibraryExW with LOAD_LIBRARY_SEARCH_SYSTEM32 is not
+  // sufficient when "icu.dll" is already loaded from somewhere other than
+  // the system directory. This is why we must pass a full path here.
+  const HMODULE icu_dll = ::LoadLibraryW(icu_dll_path.c_str());
+  if (icu_dll == nullptr) {
+    g_unavailable.store(true, std::memory_order_relaxed);
+    return Unavailable();
+  }
+
+  const auto ucal_getHostTimeZoneRef =
+      AsProcAddress<ucal_getHostTimeZone_func>(icu_dll, "ucal_getHostTimeZone");
+  const auto ucal_getTimeZoneIDForWindowsIDRef =
+      AsProcAddress<ucal_getTimeZoneIDForWindowsID_func>(
+          icu_dll, "ucal_getTimeZoneIDForWindowsID");
+
+  // Note: ucal_getHostTimeZone can be unavailable on older Windows.
+  if (!ucal_getTimeZoneIDForWindowsIDRef) {
+    g_unavailable.store(true, std::memory_order_relaxed);
+    return Unavailable();
+  }
+
+  // Store the function pointers
+  g_ucal_getHostTimeZone.store(ucal_getHostTimeZoneRef,
+                               std::memory_order_relaxed);
+  g_ucal_getTimeZoneIDForWindowsID.store(ucal_getTimeZoneIDForWindowsIDRef,
+                                         std::memory_order_relaxed);
+
+  return {true, ucal_getHostTimeZoneRef, ucal_getTimeZoneIDForWindowsIDRef};
+}
+
+// Convert wchar_t array (UTF-16) to UTF-8 string
+std::string Utf16ToUtf8(const wchar_t* ptr, size_t size) {
+  if (size > std::numeric_limits<int>::max()) {
+    return std::string();
+  }
+  const int chars_len = static_cast<int>(size);
+  int num_bytes_in_utf8 = 0;
+  {
+    std::array<char, 32> buffer;
+    num_bytes_in_utf8 =
+        ::WideCharToMultiByte(CP_UTF8, WC_ERR_INVALID_CHARS, ptr, chars_len,
```

Comment:
My reading on this function when `buffer.size() != 0`, as it always is here, is that it returns the number of bytes written to `buffer.data()`, not the number of bytes it would require to do the full job.  The latter is only returned when the output buffer size is zero.